### ISSUE TITLE
Can we use the new app directory with Nextjs 13?

### ISSUE THREAD

**Issue by algoflows** [nandorojo/solito#383]
Can we use the new app directory with Nextjs 13?

**Comment by nandorojo** [comment]
not at the moment. will revisit once it’s stable.

**Comment by algoflows** [comment]
> not at the moment. will revisit once it’s stablei think it's just been released as stable as of today... Exciting times :)

**Comment by nandorojo** [comment]
😅

**Comment by nandorojo** [comment]
is there something in particular you want to use? RNW won’t currently work with server components

**Comment by sgrund14** [comment]
@nandorojo would be nice for the new instant loading states and layout file structure on web. tho not sure if layouts can be used without server components..

https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming
https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts

plus a handful of other features, metadata handlers, intercepting routes, getServerSideProps in layout components

**Comment by dohomi** [comment]
I think a first version to add app directory support is passing a `appDir` prop to any of `<Link/>,<MotiLink/>,useRouter...` to enable a different import path of NextJS. 
```ts
// app folder
import { useRouter } from 'next/navigation';


/ / pages folder:
import { useRouter } from 'next/router';
```

**Comment by nandorojo** [comment]
are you planning on doing `use client` for everything? because RNW doesn't currently support any server components.

**Comment by nandorojo** [comment]
also the async component code won't work across platforms since expo doesn't support RSC yet

**Comment by dohomi** [comment]
I think for all router related components `use client` is also suggested inside of the app directory docs - https://nextjs.org/docs/app/api-reference/functions/use-router

I honestly think - for a NextJS and Expo app within one mono-repo - everything inside of the shared components must be used with `use client`. RSC I think would be only useful for the data fetching part - and probably for meta data and layout structure.

**Comment by dohomi** [comment]
Most of the UI libraries will definitely require `use client` at least how I see it at the moment (using Tamagui)

**Comment by nandorojo** [comment]
useParam and useParams won’t be backwards compatible with the app router, since useSearchParams and useParams are split up in the app router. Using the former turns its entire tree into client components, I believe. So I don’t want to simply throw it in every hook. 

We’ll likely need to split these into new hooks. Or I’ll add backwards compatibility into the existing hooks and then offer new ones. Not sure.

**Comment by dohomi** [comment]
probably the best is providing the appDir hooks and components as new imports. I think thats how current NextJS users have to upgrade existing codebases anyway, scanning all imports and adjusting the api and changing the imports.

**Comment by Gabrola** [comment]
With Next.js 13.4 now released with a stable App Router, it might make sense to revisit this...

> are you planning on doing `use client` for everything? because RNW doesn't currently support any server components.

This seems like a sticky situation... based on [this discussion](https://github.com/necolas/react-native-web/discussions/2474#discussioncomment-4764012), seems that the RNW team doesn't see much merit in using RSC since you can already SSR the client components, which makes sense to me. I think in most mobile applications, there will be very little non-interactive components that can become server components to be totally removed from the client bundle.

However, it seems like Next.js's direction was to make SSR and RSC mutually exclusive, which a limitation they set themselves, not a limitation of React itself. So it seems like a move to the App Router spells the doom for server-side rendered components for Solito apps, which sucks because the Next.js App Router brings a lot of useful features to the table other than RSC like the new layout architecture, parallel and intercepted routes, etc.

Having the ability to do SSR was nice... it's not a deal breaker for us, but our app benefited from a good bump in performance when using it. @nandorojo what do you think of this mess? I can see on BeatGig that you're essentially not using SSR beyond the first couple of Provider components.

Disregarding all of the above, it's clear that either way, if Solito shall support the App Router, it will need to support both the Pages Router and App Router as separate imports in parallel like @dohomi said, otherwise there will be no way to do incremental adoption.

**Comment by andresgutgon** [comment]
So if I understand it right you can mix client and RSC 

```
<Feed>
  - <Button />
  - <List />
```
Where `<Feed />` and `<List />` are RSC but `<Button />` is this:
``` javascript
// ./src/design-system/atoms/Button/index.tsx
'use client'

export default Button ({ children }) {
  const [count, setCount] = useState(0)
  return <button role='button'>{children}</button>
}
```

If I'm understanding this ok what's then the issue with Solito supporting `/app`? 

Again here to learn and super grateful for your work! 🙌

**Comment by Gabrola** [comment]
> Doing my best as a library author to keep up lol

Yeah, take it easy on yourself, you're killing it anyway 💪

**Comment by algoflows** [comment]
Hey man sick update, thanks for putting this together!!

**Comment by nandorojo** [comment]
PR opened here: #379

I'm still unsure about how injecting RNW's styles works with the App Directory. Hoping for an answer: https://twitter.com/FernandoTheRojo/status/1664325982361210881?s=20

**Comment by Gabrola** [comment]
Btw, I played around with the App Router a bit, and it seems that client components are rendered on server as well. It's just that you no longer have/need `getInitialProps`, `getServerSideProps`, etc.

**Comment by nandorojo** [comment]
yep. server components would do any data fetching of that kind.

**Comment by nandorojo** [comment]
client components should probably be called mixed components? the naming makes no sense to me

**Comment by dohomi** [comment]
> PR opened here: #397
> 
> I'm still unsure about how injecting RNW's styles works with the App Directory. Hoping for an answer: https://twitter.com/FernandoTheRojo/status/1664325982361210881?s=20

I answered on Twitter but maybe here as well. There has been efforts in Tamagui starter to support the app directory and they came up with following approach:
https://github.com/tamagui/tamagui/blob/master/starters/next-expo-solito/apps/next/app/TamaguiProvider.tsx#L30

**Comment by iquabius** [comment]
Server components are server-only, they are never rendered in the client.
Client components can be rendered on the client, but they can also be rendered on the server, the same way they always could since SSR is a thing.

When we `"use client"`, it means let this component also be rendered in the browser. So we can use state, context, event handlers, etc.

Some people have argued it should be `"use interactive"` instead.

This naming is not great indeed.

**Comment by nandorojo** [comment]
Yeah…the naming is quite unintuitive.

For those just arriving free to try the canary for v4 with App Dir support at #397.

**Comment by jguddas** [comment]
> Why? Usually a next app is one application with its own layout and ios is a expo app also with its own layout. I think is fine to have different layouts by application.

You can still have different layouts, but you would have the same routes, currently you have to manually keep the folder structure in sync.

Then you can start simple and once you want to add a custom layout for your app you just drop a `layout.native.tsx` or whatever file in the same folder.

**Comment by andresgutgon** [comment]
That would be nice in terms of dryness but I think that's out of the scope of solito. Code under nextjs app use next router and expo use expo router. I'm not sure is easy to unify the way you want but maybe yes. I don't really know

**Comment by nandorojo** [comment]
Solito v4 is now stable and supports the app directory! So I am closing this.

```sh
yarn add solito
```

https://solito.dev/v4
